### PR TITLE
[HOPSWORKS-751] Add support for serving sklearn models

### PR DESCRIPTION
**Pull request by Limmen**
- Add the option to choose sklearn in the UI when serving
- Add launcher scripts for starting and monitoring sklearn serving instances
- Add tests for sklearn serving
- Change database schema and serving user

## Make sure there is no duplicate PR for this issue

* **Please check if the PR meets the following requirements**
- [x] Adds tests for the submitted changes (for bug fixes & features)
- [ ] Passes the tests
- [x] HOPSWORKS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit

* **Post a link to the associated JIRA issue**
https://logicalclocks.atlassian.net/browse/HOPSWORKS-751

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature, support serving of SkLearn models on Hopsworks

* **What is the new behavior (if this is a feature change)?**
User can now select between tensorflow and sklearn models when serving

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
Nothing, the API is the same

* **Other information**:
Dependent PRs:

https://github.com/logicalclocks/hops-hadoop-chef/pull/141
https://github.com/logicalclocks/hops-examples/pull/55
https://github.com/logicalclocks/tensorflow-chef/pull/63
https://github.com/logicalclocks/hopsworks-chef/pull/280
https://github.com/logicalclocks/hopslog-chef/pull/16
https://github.com/logicalclocks/karamel-chef/pull/66
https://github.com/logicalclocks/hops-util-py/pull/73
https://github.com/logicalclocks/hops-docs/pull/71

Tests: https://github.com/logicalclocks/hops-testing/pull/385

### CODE REVIEW

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/InferenceController.java [hunk 1]
```diff
@@ -33,39 +35,71 @@
...
                       String verb, String inferenceRequestJson) throws InferenceException {
 
-    TfServing tfServing = tfServingFacade.findByProjectModelName(project, modelName);
-    if (tfServing == null) {
+    Serving serving = servingFacade.findByProjectAndName(project, modelName);
+    if (serving == null) {
       throw new InferenceException(RESTCodes.InferenceErrorCode.SERVING_NOT_FOUND, Level.FINE, "name: " + modelName);
     }
+  
+    if (serving.getLocalPid().equals(PID_STOPPED)) {
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/InferenceController.java]: In the ee edition we don't have the pid - this class is used by both

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/LocalhostSkLearnInferenceController.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+  CloseableHttpClient httpClient = null;
+  PoolingHttpClientConnectionManager cm = null;
+  
+  /**
+   * Initialize the HTTP connection pool configuration
+   */
+  @PostConstruct
+  public void init() {
+    cm = new PoolingHttpClientConnectionManager(30l, TimeUnit.SECONDS);
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/LocalhostSkLearnInferenceController.java]: I don't think that SkLearn and TfServing should have 2 different connection pools. The `HttpClient` is thread safe. Ideally we should have a single HttpClient for the whole application, but that's another topic. 
SkLearn and TfServing should share the HttpClient/Connection Pool. This also has the benefit that we don't need to duplicate the parameters (pool size, max connections and so on...)

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/logger/KafkaInferenceLog.java [hunk 3]
```diff
@@ -35,11 +35,11 @@
   public KafkaInferenceLog() {
   }
 
-  public KafkaInferenceLog(TfServing tfServing, String inferenceRequest,
+  public KafkaInferenceLog(Serving serving, String inferenceRequest,
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/logger/KafkaInferenceLog.java]: I think here we should log also the model type.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/LocalhostServingController.java [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+   * to the serving type.
+   *
+   * @param servingWrapper contains the user-data to validate
+   * @param project the project where the serving resides
+   * @throws ServingException
+   */
+  @Override
+  public void validateUserInput(ServingWrapper servingWrapper, Project project) throws ServingException {
+    // Check that the modelName is present
+    if (Strings.isNullOrEmpty(servingWrapper.getServing().getName())) {
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/LocalhostServingController.java]: Any way we can avoid code duplication between the LocalhostServingController and the K8sServingController?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+   * Validates user data for creating or updating a SkLearn Serving Instance
+   *
+   * @param servingWrapper the user data
+   * @param project the project to create the serving for
+   * @throws ServingException if the python environment is not activated for the project
+   */
+  public void validateUserInput(ServingWrapper servingWrapper, Project project) throws ServingException {
+
+    // Check that serving name is provided
+    if (Strings.isNullOrEmpty(servingWrapper.getServing().getName())) {
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java]: You are already checking that the name is not null in the `LocalhostServingController`

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    String scriptName = Utils.getFileName(servingWrapper.getServing().getArtifactPath());
+    if(!scriptName.contains(".py")){
+      throw new IllegalArgumentException("Script name should be a valid python script name");
+    }
+    String hdfsPath = servingWrapper.getServing().getArtifactPath();
+    if (!hdfsPath.substring(0, 7).equalsIgnoreCase("hdfs://")) {
+      hdfsPath = "hdfs://" + hdfsPath;
+    }
+    if(!servingWrapper.getServing().getArtifactPath().contains(".py")){
+      throw new IllegalArgumentException("Script name should be a valid python script name");
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java]: This is a duplicate of line 213 above

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/tf/LocalhostTfServingController.java [hunk 7]
```diff
@@ -381,55 +224,78 @@ private void startTfServingInstance(Project project, Users user, TfServing tfSer
...
+      throw new IllegalArgumentException("Batching is null");
     }
   }
 
-  private void restartTfServingInstance(Project project, Users user, TfServing currentInstance,
-                                        TfServing newInstance) throws TfServingException {
-    // Kill current TfServing instance
-    killTfServingInstance(project, currentInstance, false);
+  private void validateModelPath(String path, Integer version) throws IllegalArgumentException {
+    try {
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/tf/LocalhostTfServingController.java]: Same thing here. Any way we can avoid code duplication between the Localhost and K8s version?

File: hopsworks-common/src/main/resources/META-INF/beans.xml [hunk 8]
```diff
@@ -45,7 +45,10 @@
                       http://xmlns.jcp.org/xml/ns/javaee/beans_1_1.xsd"
         bean-discovery-mode="annotated">
     <alternatives>
+        <class>io.hops.hopsworks.common.serving.LocalhostServingController</class>
         <class>io.hops.hopsworks.common.serving.tf.LocalhostTfServingController</class>
```
Comment (SirOibaf) [inline hopsworks-common/src/main/resources/META-INF/beans.xml]: We don't need this one - it's not an alternative

File: hopsworks-common/src/main/resources/META-INF/beans.xml [hunk 9]
```diff
@@ -45,7 +45,10 @@
...
+        <class>io.hops.hopsworks.common.serving.sklearn.LocalhostSkLearnServingController</class>
```
Comment (SirOibaf) [inline hopsworks-common/src/main/resources/META-INF/beans.xml]: We don't need this one here - it's not an alternative

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/tf/LocalhostTfServingController.java [hunk 10]
```diff
@@ -17,268 +17,95 @@
...
 import static io.hops.hopsworks.common.hdfs.HdfsUsersController.USER_NAME_DELIMITER;
+import static io.hops.hopsworks.common.serving.LocalhostServingController.PID_STOPPED;
+import static io.hops.hopsworks.common.serving.LocalhostServingController.SERVING_DIRS;
 
+/**
+ * Localhost Tensorflow Serving Controller
+ *
+ * Launches/Kills a local tensorflow-serving-server for serving tensorflow Models
+ */
 @Alternative
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/tf/LocalhostTfServingController.java]: This is not an `alternative` anymore. `alternative` is used when there are multiple classes implementing the same interface and glassfish needs to know which one to use.
This is not an interface anymore

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+import static io.hops.hopsworks.common.hdfs.HdfsUsersController.USER_NAME_DELIMITER;
+import static io.hops.hopsworks.common.serving.LocalhostServingController.PID_STOPPED;
+import static io.hops.hopsworks.common.serving.LocalhostServingController.SERVING_DIRS;
+
+/**
+ * Localhost SkLearn Serving Controller
+ *
+ * Launches/Kills a local Flask Server for Serving SkLearn Models
+ */
+@Alternative
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java]: Same as for the `LocalhostTfServingController` - This class does not implement any interface, so it should not be marked as `alternative`

**Comment by tkakantousis** [comment]
@Limmen can you remove this file? hopsworks-api/src/main/java/io/hops/hopsworks/api/project/ProjectService.java.orig

File: hopsworks-IT/src/test/ruby/spec/inference_spec.rb [hunk 12]
```diff
@@ -174,7 +174,7 @@
           end
 
           it "should receive an error if the input payload is empty" do
-            post "#{ENV['HOPSWORKS_API']}/project/#{@project[:id]}/inference/models/#{@serving[:model_name]}:predict"
+            post "#{ENV['HOPSWORKS_API']}/project/#{@project[:id]}/inference/models/#{@serving[:name]}:predict"
             expect_json(errorCode: 250008)
             expect_status(400)
           end
```
[the remark this answers is not in the document]
Comment (Limmen) [reply hopsworks-IT/src/test/ruby/spec/inference_spec.rb]: Done

File: hopsworks-IT/src/test/ruby/spec/serving_spec.rb [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+ DAMAGES OR  OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+=end
+
+require 'json'
+
+describe "On #{ENV['OS']}" do
+  describe 'tfserving' do
+    before (:all) do
+      if ENV['OS'] == "centos"
```
[the remark this answers is not in the document]
Comment (Limmen) [reply hopsworks-IT/src/test/ruby/spec/serving_spec.rb]: Done

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/serving/ServingView.java [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+  public TopicDTO getKafkaTopicDTO() {
+    return kafkaTopicDTO;
+  }
+
+  public void setKafkaTopicDTO(TopicDTO kafkaTopicDTO) {
+    this.kafkaTopicDTO = kafkaTopicDTO;
+  }
+  
+  @ApiModelProperty(value = "Type of serving, sklearn or tfserving")
+  public Integer getServingType() {
```
[the remark this answers is not in the document]
Comment (Limmen) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/serving/ServingView.java]: Done

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/serving/ServingView.java [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+    return servingType;
+  }
+
+  public ServingType getServingTypeEnum() {
```
[the remark this answers is not in the document]
Comment (Limmen) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/serving/ServingView.java]: Done

File: hopsworks-common/src/main/resources/META-INF/beans.xml (hunk 8 above)
Comment (Limmen) [reply hopsworks-common/src/main/resources/META-INF/beans.xml]: Done

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/tf/LocalhostTfServingController.java (hunk 10 above)
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/tf/LocalhostTfServingController.java]: Done

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java (hunk 11 above)
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java]: Done

**Comment by Limmen** [comment]
@tkakantousis Done

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/serving/ServingView.java [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+  private Integer id;
+  private String name;
+  private String artifactPath;
+  private Integer modelVersion;
+  private Integer availableInstances;
+  private Integer requestedInstances;
+  private Integer nodePort;
+  private Date created;
+  private Boolean batchingEnabled;
+  private Integer servingType;
```
[the remark this answers is not in the document]
Comment (Limmen) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/serving/ServingView.java]: Done

File: hopsworks-common/src/main/resources/META-INF/beans.xml (hunk 9 above)
Comment (Limmen) [reply hopsworks-common/src/main/resources/META-INF/beans.xml]: Done

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/serving/ServingFacade.java [hunk 17]
```diff
@@ -59,74 +59,78 @@ protected EntityManager getEntityManager() {
...
     }
 
-    if (newTfServing.isBatchingEnabled() != null) {
-      dbTfServing.setBatchingEnabled(newTfServing.isBatchingEnabled());
+    if (newServing.isBatchingEnabled() != null) {
+      dbServing.setBatchingEnabled(newServing.isBatchingEnabled());
     }
 
-    return merge(dbTfServing);
+    if (newServing.getServingType() != null) {
```
[the remark this answers is not in the document]
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/serving/ServingFacade.java]: Done

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/security/CertificateMaterializer.java [hunk 18]
```diff
@@ -1136,7 +1137,47 @@ private String normalizeURI(String uri) {
...
+  /**
+   * Returns the path to materialized truststore for a project-user
+   *
+   * @param project the project for which the truststore will be used
+   * @param user    the user for which the truststore will be used
+   * @return path to truststore
+   */
+  public String getUserTransientTruststorePath(Project project, Users user) {
+    String transientDir = settings.getHopsworksTmpCertDir();
+    String truststoreName = project.getName() + "__" + user.getUsername() + Settings.TRUSTSTORE_SUFFIX;
```
[the remark this answers is not in the document]
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/security/CertificateMaterializer.java]: Done

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/InferenceController.java (hunk 1 above)
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/InferenceController.java]: Done

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/LocalhostSkLearnInferenceController.java (hunk 2 above)
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/LocalhostSkLearnInferenceController.java]: Done

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/logger/KafkaInferenceLog.java (hunk 3 above)
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/inference/logger/KafkaInferenceLog.java]: Done

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java (hunk 5 above)
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java]: Done

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java (hunk 6 above)
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/sklearn/LocalhostSkLearnServingController.java]: Done

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/tf/LocalhostTfServingController.java (hunk 7 above)
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/tf/LocalhostTfServingController.java]: Done, moved common stuff to a Util Bean

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/LocalhostServingController.java (hunk 4 above)
Comment (Limmen) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/LocalhostServingController.java]: Done, moved common stuff to a Util Bean

**Comment by Limmen** [comment]
> @Limmen can you remove thes files?
> 
> 1. hopsworks-api/src/main/java/io/hops/hopsworks/api/project/ProjectService.java.orig
> 2. hopsworks-api/src/main/java/io/hops/hopsworks/rest/application/config/ApplicationConfig.java.orig
> 3. hopsworks-common/src/main/java/io/hops/hopsworks/common/project/ProjectController.java.orig

Done

**Pull request merged**